File: internal/api/editor_mode_resolver_test.go
package api

import (
	"testing"

	"github.com/stretchr/testify/require"
)

func TestResolveEditorForPathMapsKnownExtensions(t *testing.T) {
	cases := []struct {
		path       string
		mode       editorMode
		inline     bool
		syntax     bool
		image      bool
		markdownUI bool
	}{
		{path: "/posts/2026-02-06-launch.md", mode: editorModeMarkdown, inline: true, markdownUI: true},
		{path: "/notes/quick.txt", mode: editorModeText},
		{path: "/notes/main.go", mode: editorModeCode, syntax: true},
		{path: "/assets/mockup.png", mode: editorModeImage, image: true},
	}

	for _, tc := range cases {
		resolution := resolveEditorForPath(tc.path)
		require.Equal(t, tc.mode, resolution.Mode, tc.path)
		require.Equal(t, tc.inline, resolution.Capabilities.SupportsInlineComments, tc.path)
		require.Equal(t, tc.syntax, resolution.Capabilities.SupportsSyntaxHighlight, tc.path)
		require.Equal(t, tc.image, resolution.Capabilities.SupportsImagePreview, tc.path)
		require.Equal(t, tc.markdownUI, resolution.Capabilities.SupportsMarkdownView, tc.path)
	}
}

func TestResolveEditorForPathUnknownExtensionFallsBackToSafeTextMode(t *testing.T) {
	resolution := resolveEditorForPath("/notes/spec.custom")
	require.Equal(t, editorModeText, resolution.Mode)
	require.True(t, resolution.Capabilities.Editable)
	require.False(t, resolution.Capabilities.SupportsInlineComments)
	require.False(t, resolution.Capabilities.SupportsSyntaxHighlight)
	require.False(t, resolution.Capabilities.SupportsImagePreview)
}

File: web/src/components/content-review/editorModeResolver.test.ts
import { describe, expect, it } from "vitest";
import { resolveEditorComponent, resolveEditorForPath } from "./editorModeResolver";

describe("editor mode resolver", () => {
  it("maps known extensions to expected editor modes", () => {
    expect(resolveEditorForPath("/posts/2026-02-06-launch.md").editorMode).toBe("markdown");
    expect(resolveEditorForPath("/notes/scratch.txt").editorMode).toBe("text");
    expect(resolveEditorForPath("/notes/main.ts").editorMode).toBe("code");
    expect(resolveEditorForPath("/assets/mockup.PNG").editorMode).toBe("image");
  });

  it("maps unknown extensions to safe text fallback", () => {
    const resolution = resolveEditorForPath("/notes/spec.custom");
    expect(resolution.editorMode).toBe("text");
    expect(resolution.capabilities.supportsInlineComments).toBe(false);
    expect(resolution.capabilities.supportsSyntaxHighlight).toBe(false);
    expect(resolution.capabilities.supportsImagePreview).toBe(false);
  });

  it("drives editor component selection deterministically", () => {
    expect(resolveEditorComponent("/posts/2026-02-06-launch.md")).toBe("markdown_review");
    expect(resolveEditorComponent("/notes/scratch.txt")).toBe("plain_text");
    expect(resolveEditorComponent("/notes/server.go")).toBe("code_editor");
    expect(resolveEditorComponent("/assets/diagram.gif")).toBe("image_preview");
    expect(resolveEditorComponent("/notes/unknown.ext")).toBe("plain_text");
  });
});
